### PR TITLE
[Perf] [cuda] Automatically use read-only data cache load

### PR DESCRIPTION
**Pull request by yuanming-hu**
About read-only data cache on CUDA: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#ldg-function

Note that on newer GPUs the performance seems less pronounced compared to old ones. For example, on `stable_fluid.py`
- on RTX 3090, this optimization leads to no performance gain
- on GTX 1080, this optimization leads to a ~4% performance gain. Compared to the original `stable_fluid.py` with manual hints, the compiler-optimized version is 1% faster. (So I removed the manual hints.)

Also, note that the backend compiler may be able to do this optimization automatically once it finds a memory load is read-only. We are doing it just to enforce the use of `__ldg` instructions. Taichi is able to do this since the `SNode` system is very friendly to aliasing analysis.

Anyway, we can remove the manual `ti.cache_read_only` hints in `stable_fluid.py` since the compiler can do it automatically now, and a 4% automatic performance boost is not bad. I believe there are other cases where read-only caches play a bigger role (I've seen cases where `__ldg` makes the program run 2x faster), but for now, we are just using `stable_fluid` for a quick benchmark.

I also introduced `SNodeAccessFlag`, as a more systematic solution compared to previous `int`-based access flagging.

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: python/taichi/lang/__init__.py [hunk 1]
```diff
@@ -310,7 +311,6 @@ def visit(node):
     visit(ti.root)
 
 
-schedules = [parallelize, vectorize, block_dim, cache]
```
Comment (yuanming-hu) [inline python/taichi/lang/__init__.py:313]: Turns out to be never used...

**Comment by yuanming-hu** [comment]
This seems to fail on the build bots - let me take another look and re-request for review when it's ready. Thanks!

**Review by archibate — APPROVED** [review]
Cool! Any good I can help for opengl as well?

**Comment by yuanming-hu** [comment]
> Cool! Any good I can help for opengl as well?

Thanks! I'm not sure if read-only data cache loads are available to OpenGL. A related thing in OpenGL is probably texture memory: https://www.sciencedirect.com/topics/computer-science/texture-memory (Not sure if that is programmable though)

**Pull request merged**